### PR TITLE
BigQuery: Add dry_run option to BigQuery magic

### PR DESCRIPTION
**Pull request by shubha-rajan**
See #8143
Adding the `--dry_run` flag returns a `QueryJob` instead of a pandas `DataFrame` as requested in the original issue. The `QueryJob` can also be stored in a variable if the `destination_var` magic argument is present

### CODE REVIEW

File: bigquery/google/cloud/bigquery/magics.py [hunk 1]
```diff
@@ -324,6 +324,15 @@ def _run_query(client, query, job_config=None):
         "the context default_query_job_config.maximum_bytes_billed."
     ),
 )
+@magic_arguments.argument(
+    "--dry_run",
+    action="store_true",
+    default=False,
+    help=(
+        "Sets query to be a dry run to estimate costs"
```
Comment (plamut) [inline bigquery/google/cloud/bigquery/magics.py]: A period and a space are missing at the end of the sentence, looks strange in the help output.

```suggestion
        "Sets query to be a dry run to estimate costs. "
```

**Comment by shubha-rajan** [comment]
I'll investigate that error being printed out.  I agree that it would be confusing to users, and don't think it's outside of the scope of the PR

**Comment by tswast** [comment]
When you run a dry run query, a job is not actually created, so fetching the results fails.

**Comment by tswast** [comment]
Have you run this in a notebook, yet? I'm curious what the output looks like. We probably want to improve the `__str__` or `__repr__` functions on `QueryJob` to at least show [`estimated_bytes_processed`](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJob.html#google.cloud.bigquery.job.QueryJob.estimated_bytes_processed), because that's what people running dryrun queries are going to care about.

**Comment by shubha-rajan** [comment]
@tswast as of now, running it in a notebook is silent. I tried modifying `_run_query()` and that made the error @plamut was referring to go away. I think the next step is to look at modifying those functions in `QueryJob` 
image

**Comment by shubha-rajan** [comment]
6efa408 allows the cell to output the total bytes processed. adding a variable argument allows the user to inspect values the returned `QueryJob`'s properties including `total_bytes_processed` and `total_bytes_billed`
image

**Comment by shubha-rajan** [comment]
So currently, `destination_var` doesn't return anything if an error occurs by default, even when the `--dry_run` flag isn't present. I investigated adding in that functionality, but the problem is that if calling `_run_query` raises an exception, it won't return a `QueryJob`. I suspect getting around that would mean making changes to exception handling in `QueryJob` and/or `Client` and idk what might end up breaking because of that 😅 

As for printing out the SQL queries along with error messages, that's working as of the last commit. 
image

File: bigquery/google/cloud/bigquery/job.py [hunk 2]
```diff
@@ -2876,6 +2876,29 @@ def _format_for_exception(query, job_id):
 
         return template.format(job_id=job_id, header=header, ruler=ruler, body=body)
 
+    def _begin(self, client=None, retry=DEFAULT_RETRY):
+        """API call:  begin the job via a POST request
+
+        See
+        https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/insert
+
+        :type client: :class:`~google.cloud.bigquery.client.Client` or
```
Comment (tswast) [inline bigquery/google/cloud/bigquery/job.py]: Nit: We're using Google-style docstrings for new functions.

```suggestion
        Args:
            client (Optional[google.cloud.bigquery.client.Client]):
                The client to use. If not passed, falls back to the ``client``
                associated with the job object.
```

Style guide: http://google.github.io/styleguide/pyguide.html#383-functions-and-methods (We follow PEP-8 for everything except docstrings, though.)

See examples at https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google

File: bigquery/google/cloud/bigquery/job.py [hunk 3]
```diff
@@ -2876,6 +2876,29 @@ def _format_for_exception(query, job_id):
...
+                    ``NoneType``
+        :param client: the client to use.  If not passed, falls back to the
+                    ``client`` stored on the current dataset.
```
Comment (tswast) [inline bigquery/google/cloud/bigquery/job.py]: Dataset objects don't actually have `client` properties anymore, and haven't for a couple years. I've suggested some alternative language above. We should probably fix the docstring this was copied from, too.

File: bigquery/tests/unit/test_magics.py [hunk 4]
```diff
@@ -626,6 +645,107 @@ def test_bigquery_magic_without_bqstorage(monkeypatch):
...
+    )
+    query_job_mock = mock.create_autospec(
+        google.cloud.bigquery.job.QueryJob, instance=True
+    )
+    run_query_patch = mock.patch(
+        "google.cloud.bigquery.magics._run_query", autospec=True
+    )
+
+    sql = "SELECT 17 AS num"
+    result = pandas.DataFrame([17], columns=["num"])
```
Comment (tswast) [inline bigquery/tests/unit/test_magics.py]: We don't need this `result` variable or the mock of `to_dataframe`, correct? We should be able to remove this from the dry run unit tests.

File: bigquery/tests/unit/test_magics.py [hunk 5]
```diff
@@ -626,6 +645,107 @@ def test_bigquery_magic_without_bqstorage(monkeypatch):
...
```
Comment (tswast) [inline bigquery/tests/unit/test_magics.py]: Ditto. We should be able to remove `result` from the dry run unit tests.

File: bigquery/google/cloud/bigquery/job.py (hunk 2 above)
Comment (plamut) [reply bigquery/google/cloud/bigquery/job.py]: I also noticed that, but then realized the method was based on one of the existing methods, and that method also uses Sphinx-style docstrings.

If changing the docstring to Google style, I suggest doing the same in the original parent class's method as well.

Comment (plamut) [reply bigquery/google/cloud/bigquery/job.py]: Ticket - https://github.com/googleapis/google-cloud-python/issues/9092

File: bigquery/tests/unit/test_magics.py (hunk 5 above)
Comment (shubha-rajan) [reply bigquery/tests/unit/test_magics.py]: removed

File: bigquery/tests/unit/test_magics.py (hunk 4 above)
Comment (shubha-rajan) [reply bigquery/tests/unit/test_magics.py]: removed

**Pull request merged**